index.bs

```diff
@@ -12337,18 +12336,73 @@ The value of the [=function object=]’s <code class="idl">name</code> property
 
 <h3 id="es-platform-objects">Platform objects implementing interfaces</h3>
 
+<div algorithm>
+    An ECMAScript value |value| <dfn id="es-is-platform-object">is a platform object</dfn>
+    if [$Type$](|value|) is Object and if |value| has a \[[PrimaryInterface]] internal slot.
+</div>
+
+<div algorithm>
+    An ECMAScript value |value| <dfn export>implements</dfn> an [=interface=] |interface|
+    if |value| [=is a platform object=] and one of the following is true:
+    <ul>
+        <li>|value|.\[[PrimaryInterface]] is |interface|.</li>
+        <li>The set of [=inherited interfaces=] of |value|.\[[PrimaryInterface]] contains |interface|.</li>
```

Comment:
Theoretically it'd be better if these inherited interfaces were stored directly on the object, no?

Reply:
Why's that? I thought it'd be better if the data structures were "normalized"--this makes it more clear to implementers that they don't need to denormalize either.

Reply:
Because "inherited interfaces" as currently defined is global. To me it seems better to not rely on a global lookup for something like this. Note that inlining this would make this [[PlatformBrand]].

Reply:
Right, this PR is trying to get at the same thing as you were proposing with `[[PlatformBrand]]`.

I don't see the downside to the implementation or specification strategy to use shared data structures here during the actual check itself. We will have to look up the brand set from these data structures at some point one way or another, just a question of whether we shift the lookup a bit earlier.

Reply:
One thing we could do is explain what an interface is in terms of internal slots, so it doesn't sound like we're searching through all the interfaces to find the set of inherited interfaces, or "cache" the set of inherited interfaces. But this seems like excessive formalism for no particular purpose--it's pretty clear how to represent these with efficient data structures.

Reply:
I guess as long as the "API" for other standards is clear it doesn't matter much how it's implemented here.

Implementation-wise, I know Gecko stores the set of interfaces on the object itself, do Chromium and WebKit do something else? If they all do that, what is the reason to not do it here and move the location where we invoke "inherited interfaces"? All things being equal I'd much rather implicitly recommend an optimal implementation strategy.

Reply:
To clarify, what I like about this form of specification is that it makes it clear that, even when an object is in a deoptimized hashtable mode, no more than constant space is needed to represent the set of platform brands needed for a particular instance.

Reply:
How does that avoid the brand being per Realm? It seems the "function template" would have to be scoped to an Agent, no?

Reply:
Yep, I believe that's how it works.

Reply:
Yeah, I guess it's cached in [GetWrapperImpl](https://cs.chromium.org/chromium/src/gin/wrappable.cc?type=cs&q=SetObjectTemplate&g=0&l=42) isolate-wide (i.e., per agent).